lib/marisa/grimoire/vector/vector.h

```diff
@@ -231,8 +235,19 @@ class Vector {
     MARISA_DEBUG_IF(new_buf.get() == NULL, MARISA_MEMORY_ERROR);
     T *new_objs = reinterpret_cast<T *>(new_buf.get());
 
-    for (std::size_t i = 0; i < size_; ++i) {
-      new (&new_objs[i]) T(objs_[i]);
+#if __cplusplus >= 201103L
+    constexpr bool trivially_copyable = std::is_trivially_copyable<T>::value;
+#else
+    const bool trivially_copyable = false;
+#endif
+    if (trivially_copyable) {
+      std::memcpy(reinterpret_cast<void*>(new_objs),
```

Comment:
This can be `memmove`.

I was just playing around with godbolt.org and noticed that `std::copy_n`, `uninitialized_copy_n` and `uninitialized_move_n` all compile to `memmove` for `char`. This definitely makes the `uninitialized_move_n` plus compatibility wrapper seem nicer.

Reply:
`memcpy` is usually faster than `memmove` (but requires the memory regions to not overlap).

I'm surprised about `std::copy_n`, it should compile to `memcpy` as it also forbids overlap.

Reply:
Ah, I just assumed that the optimizer was doing the most efficient thing, since it usually does, so got `memcpy` and `memmove` switched in my head.

My initial example was actually insufficient, because my pointers could have overlapped. When one comes from `new`, `memcpy` is used.

https://godbolt.org/z/HZ5VU5